In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
%matplotlib inline
import seaborn as sns

In [2]:
final=pd.read_csv('Final_Data.csv')

In [3]:
def diabetes(df):
    c=0
    for i in range(len(df)):
        if(( df.iloc[i,1].startswith('250.') == True) or ( df.iloc[i,2].startswith('250.') == True) or ( df.iloc[i,3].startswith('250.') == True)):
            c+=1
    return c 

In [4]:
diabetes(final)
# 20166 have diag1 or diag2 or diag3 have values 250.xx

20166

In [5]:
final[final['patient_nbr']==88785891]

patient_nbr  diag_1  diag_2  diag_3  number_diagnoses  \
38307      88785891  250.13     244     490                 3   
40252      88785891  250.11     244       ?                 2   
40661      88785891  250.13     536     V58                 6   
44515      88785891  250.11     465     244                 3   
45147      88785891  250.11     V58     V15                 9   
45986      88785891   250.1     276     276                 9   
50167      88785891       8  250.01     V09                 7   
50393      88785891  250.11     305     V58                 6   
50773      88785891  250.11     244     V15                 6   
51519      88785891  250.11     V15     V58                 4   
52057      88785891  250.11     300     296                 5   
53771      88785891  250.12     530     578                 7   
54497      88785891   250.1     244     V02                 4   
60263      88785891  250.11     540     276                 9   
62759      88785891  250.11   250.6     536                 9   
63165      88785891   250.1     305     305                 9   
65667      88785891  250.11     305     296                 4   
67968      88785891  250.11     346     244                 5   
71029      88785891   250.1     276     244                 5   
72068      88785891  250.13     599     V58                 6   
74405      88785891  250.02     244     V58                 3   
74876      88785891  250.13     305     V58                 9   
75267      88785891   250.1     276     276                 9   
76525      88785891  250.11     530     112                 8   
76907      88785891     789     787     250                 6   
78355      88785891  250.13     305     276                 9   
78882      88785891  250.13     305     242                 4   
79202      88785891  250.11     305     V58                 7   
79467      88785891  250.12     304     296                 7   
82693      88785891     780  250.03     784                 8   
87063      88785891     789  250.03     276                 9   
87854      88785891  250.11     305     V15                 9   
88392      88785891  250.11     276     276                 9   
89964      88785891     787     305  250.03                 9   
90913      88785891  250.13   250.6     536                 7   
91444      88785891     787     305     305                 9   
97433      88785891       8  250.01     276                 9   
97732      88785891  250.13     276     276                 9   
99950      88785891  250.11     276     305                 5   
101385     88785891     296     305     292                 9   

        admission_type_id  discharge_disposition_id  admission_source_id  \
38307                   1                         1                    7   
40252                   1                         1                    7   
40661                   1                         1                    7   
44515                   3                         1                    7   
45147                   1                         1                    7   
45986                   2                         1                    7   
50167                   2                         1                    7   
50393                   1                         1                    7   
50773                   2                         7                    7   
51519                   2                         1                    7   
52057                   2                         1                    7   
53771                   2                         1                    7   
54497                   2                         1                    7   
60263                   1                         1                    7   
62759                   1                         1                    7   
63165                   2                         1                    7   
65667                   1                         

In [6]:
final['encounter_id'].nunique()

101766

In [7]:
final['patient_nbr'].nunique()

71518

In [8]:
final['diag_1'].unique()

array(['250.83', '276', '648', '8', '197', '414', '428', '398', '434',
       '250.7', '157', '518', '999', '410', '682', '402', '737', '572',
       'V57', '189', '786', '427', '996', '277', '584', '462', '473',
       '411', '174', '486', '998', '511', '432', '626', '295', '196',
       '250.6', '618', '182', '845', '423', '808', '250.4', '722', '403',
       '250.11', '784', '707', '440', '151', '715', '997', '198', '564',
       '812', '38', '590', '556', '578', '250.32', '433', 'V58', '569',
       '185', '536', '255', '250.13', '599', '558', '574', '491', '560',
       '244', '250.03', '577', '730', '188', '824', '250.8', '332', '562',
       '291', '296', '510', '401', '263', '438', '70', '250.02', '493',
       '642', '625', '571', '738', '593', '250.42', '807', '456', '446',
       '575', '250.41', '820', '515', '780', '250.22', '995', '235',
       '250.82', '721', '787', '162', '724', '282', '514', 'V55', '281',
       '250.33', '530', '466', '435', '250.12', 'V53', '789', '

### Replacing ? with 'missing' in diag 1,2 and 3

In [9]:
final[final['diag_1']=='?'].shape

(21, 50)

In [10]:
final['diag_1'].replace({'?':'missing'},inplace=True)

In [11]:
final[final['diag_1']=='missing'].shape

(21, 50)

In [12]:
final[final['diag_2']=='?'].shape

(358, 50)

In [13]:
final['diag_2'].replace({'?':'missing'},inplace=True)
final[final['diag_2']=='missing'].shape

(358, 50)

In [14]:
final[final['diag_3']=='?'].shape

(1423, 50)

In [15]:
final['diag_3'].replace({'?':'missing'},inplace=True)
final[final['diag_3']=='missing'].shape

(1423, 50)

In [16]:
final['number_diagnoses'].unique() # no nulls

array([ 1,  9,  6,  7,  5,  8,  3,  4,  2, 16, 12, 13, 15, 10, 11, 14],
      dtype=int64)

In [17]:
final['admission_type_id'].unique() # no nulls

array([6, 1, 2, 3, 4, 5, 8, 7], dtype=int64)

### Replacing admission type id numbers with actual mapping to make the data nominal as it is meant to be

In [18]:
final['admission_type_id'].replace([1,2,3,4,5,6,7,8],
                                   ['Emergency','Urgent','Elective','Newborn','NA','NULL','Trauma Center','Not Mapped'],inplace=True)

In [19]:
final['admission_type_id'].replace(['Not Mapped','Trauma Center'],
                                   ['Not_Mapped','Trauma_Center'],inplace=True)

In [20]:
final['admission_type_id'].value_counts()

Emergency        53990
Elective         18869
Urgent           18480
NULL              5291
NA                4785
Not_Mapped         320
Trauma_Center       21
Newborn             10
Name: admission_type_id, dtype: int64

 ### Replacing discharge disposition id numbers with actual mapping to make the data nominal as it is meant to be

In [21]:
final['discharge_disposition_id'].unique()

array([25,  1,  3,  6,  2,  5, 11,  7, 10,  4, 14, 18,  8, 13, 12, 16, 17,
       22, 23,  9, 20, 15, 24, 28, 19, 27], dtype=int64)

In [22]:
final['discharge_disposition_id'].value_counts().sort_index()

1     60234
2      2128
3     13954
4       815
5      1184
6     12902
7       623
8       108
9        21
10        6
11     1642
12        3
13      399
14      372
15       63
16       11
17       14
18     3691
19        8
20        2
22     1993
23      412
24       48
25      989
27        5
28      139
Name: discharge_disposition_id, dtype: int64

discharge_disposition_id	description
Discharged to home
Discharged/transferred to another short term hospital
Discharged/transferred to SNF
Discharged/transferred to ICF
Discharged/transferred to another type of inpatient care institution
Discharged/transferred to home with home health service
Left AMA
Discharged/transferred to home under care of Home IV provider
Admitted as an inpatient to this hospital
Neonate discharged to another hospital for neonatal aftercare
Expired
Still patient or expected to return for outpatient services
Hospice / home
Hospice / medical facility
Discharged/transferred within this institution to Medicare approved swing bed
Discharged/transferred/referred another institution for outpatient services
Discharged/transferred/referred to this institution for outpatient services
NULL
Expired at home. Medicaid only, hospice.
Expired in a medical facility. Medicaid only, hospice.
Expired, place unknown. Medicaid only, hospice.
Discharged/transferred to another rehab fac including rehab units of a hospital .
Discharged/transferred to a long term care hospital.
Discharged/transferred to a nursing facility certified under Medicaid but not certified under Medicare.
Not Mapped
Unknown/Invalid
Discharged/transferred to a federal health care facility.
Discharged/transferred/referred to a psychiatric hospital of psychiatric distinct part unit of a hospital
Discharged/transferred to a Critical Access Hospital (CAH).
Discharged/transferred to another Type of Health Care Institution not Defined Elsewhere


In [23]:
final['discharge_disposition_id'].replace(np.arange(1,31),  ['ToHome' , 'ToShortTermHosp','ToSNF','ToICF','ToInpatientCareInst','ToHomeWithHomeHealthCare',
                                   'LeftAMA','ToHomeIVProvider','InpatientInSameHosp','AnotherHospNeonatalCare','Expired',
                                   'StillPatientOROutpatient','Hospice/Home','Hospice/Med.Facility','MedicareApprovedSwingBed',
                                   'AnotherInst.Outpatient','SameInst.Outpatient','NULL','ExpiredAtHome.MedicaidOnly',
                                  'ExpiredInMedFacility.MedicaidOnly','ExpiredUnknown.MedicaidOnly','ToRehab','ToLongTermCare',
                                   'ToNursingFac(MedicaidCertified,MedicareNotCertified)','Not_Mapped','Unknown/Invalid',
                                   'ToFederalHealthCareFac.','ToPsychiatricHosp.','ToCAHosp','ToUnknownHCHosp.']
                                   ,inplace=True)

In [24]:
final['discharge_disposition_id'].value_counts()

ToHome                                                  60234
ToSNF                                                   13954
ToHomeWithHomeHealthCare                                12902
NULL                                                     3691
ToShortTermHosp                                          2128
ToRehab                                                  1993
Expired                                                  1642
ToInpatientCareInst                                      1184
Not_Mapped                                                989
ToICF                                                     815
LeftAMA                                                   623
ToLongTermCare                                            412
Hospice/Home                                              399
Hospice/Med.Facility                                      372
ToPsychiatricHosp.                                        139
ToHomeIVProvider                                          108
Medicare

 ### Replacing admission source id numbers with actual mapping to make the data nominal as it is meant to be

In [25]:
final['admission_source_id'].value_counts()

7     57494
1     29565
17     6781
4      3187
6      2264
2      1104
5       855
3       187
20      161
9       125
8        16
22       12
10        8
11        2
14        2
25        2
13        1
Name: admission_source_id, dtype: int64

In [26]:
final['admission_source_id'].replace(np.arange(1,27),['PhysicianReferral','ClinicReferral','HMOReferral','Transferfromahospital',' TransferfromaSkilledNursingFacility(SNF)',
 'Transferfromanotherhealthcarefacility', 'EmergencyRoom',
'Court/LawEnforcement','NotAvailable',
'Transferfromcritialaccesshospital','NormalDelivery',
'PrematureDelivery', 'SickBaby',
'ExtramuralBirth','NotAvailable','','NULL',
 'TransferFromAnotherHomeHealthAgency',
'ReadmissiontoSameHomeHealthAgency',
'NotMapped','Unknown/Invalid',
'Transferfromhospitalinpt/samefacresultinasepclaim',
'Borninsidethis hospital',' Bornoutsidethishospital',
'TransferfromAmbulatorySurgeryCenter',
'TransferfromHospice'],inplace=True)

In [27]:
final['admission_source_id'].value_counts()

EmergencyRoom                                        57494
PhysicianReferral                                    29565
NULL                                                  6781
Transferfromahospital                                 3187
Transferfromanotherhealthcarefacility                 2264
ClinicReferral                                        1104
 TransferfromaSkilledNursingFacility(SNF)              855
HMOReferral                                            187
NotMapped                                              161
NotAvailable                                           125
Court/LawEnforcement                                    16
Transferfromhospitalinpt/samefacresultinasepclaim       12
Transferfromcritialaccesshospital                        8
ExtramuralBirth                                          2
TransferfromAmbulatorySurgeryCenter                      2
NormalDelivery                                           2
SickBaby                                                

In [28]:
final['medical_specialty'].value_counts()

?                                    49949
InternalMedicine                     14635
Emergency/Trauma                      7565
Family/GeneralPractice                7440
Cardiology                            5352
Surgery-General                       3099
Nephrology                            1613
Orthopedics                           1400
Orthopedics-Reconstructive            1233
Radiologist                           1140
Pulmonology                            871
Psychiatry                             854
Urology                                685
ObstetricsandGynecology                671
Surgery-Cardiovascular/Thoracic        652
Gastroenterology                       564
Surgery-Vascular                       533
Surgery-Neuro                          468
PhysicalMedicineandRehabilitation      391
Oncology                               348
Pediatrics                             254
Hematology/Oncology                    207
Neurology                              203
Pediatrics-

#### Replacing ? with missing in medical speciality

In [29]:
final['medical_specialty'].replace({'?':'missing'},inplace=True) # as close to 50% data is missing, we cannot replace with mode
final['medical_specialty'].value_counts()

missing                              49949
InternalMedicine                     14635
Emergency/Trauma                      7565
Family/GeneralPractice                7440
Cardiology                            5352
Surgery-General                       3099
Nephrology                            1613
Orthopedics                           1400
Orthopedics-Reconstructive            1233
Radiologist                           1140
Pulmonology                            871
Psychiatry                             854
Urology                                685
ObstetricsandGynecology                671
Surgery-Cardiovascular/Thoracic        652
Gastroenterology                       564
Surgery-Vascular                       533
Surgery-Neuro                          468
PhysicalMedicineandRehabilitation      391
Oncology                               348
Pediatrics                             254
Hematology/Oncology                    207
Neurology                              203
Pediatrics-

In [30]:
final['race'].value_counts()

Caucasian          76099
AfricanAmerican    19210
?                   2273
Hispanic            2037
Other               1506
Asian                641
Name: race, dtype: int64

#### Replacing ? with mode - Caucasian

In [31]:
final['race'].replace({'?':'Caucasian'},inplace=True)
final['race'].value_counts()

Caucasian          78372
AfricanAmerican    19210
Hispanic            2037
Other               1506
Asian                641
Name: race, dtype: int64

In [32]:
final['gender'].value_counts()

Female             54708
Male               47055
Unknown/Invalid        3
Name: gender, dtype: int64

In [33]:
final['gender'].replace({'Unknown/Invalid':'Female'},inplace=True)
# replacing Unknown/Invalid with mode - Female
final['gender'].value_counts()

Female    54711
Male      47055
Name: gender, dtype: int64

In [34]:
final['age'].value_counts() # no nulls

[70-80)     26068
[60-70)     22483
[50-60)     17256
[80-90)     17197
[40-50)      9685
[30-40)      3775
[90-100)     2793
[20-30)      1657
[10-20)       691
[0-10)        161
Name: age, dtype: int64

In [35]:
final['weight'].value_counts()

?            98569
[75-100)      1336
[50-75)        897
[100-125)      625
[125-150)      145
[25-50)         97
[0-25)          48
[150-175)       35
[175-200)       11
>200             3
Name: weight, dtype: int64

In [36]:
final['weight'].replace({'?':'missing'},inplace=True)
# replacing ? with missing because almost all are missing values
final['weight'].value_counts()

missing      98569
[75-100)      1336
[50-75)        897
[100-125)      625
[125-150)      145
[25-50)         97
[0-25)          48
[150-175)       35
[175-200)       11
>200             3
Name: weight, dtype: int64

In [37]:
final['payer_code'].value_counts()

?     40256
MC    32439
HM     6274
SP     5007
BC     4655
MD     3532
CP     2533
UN     2448
CM     1937
OG     1033
PO      592
DM      549
CH      146
WC      135
OT       95
MP       79
SI       55
FR        1
Name: payer_code, dtype: int64

In [38]:
final['payer_code'].replace({'?':'missing'},inplace=True)
# replacing ? with missing because many vals are missing values
final['payer_code'].value_counts()

missing    40256
MC         32439
HM          6274
SP          5007
BC          4655
MD          3532
CP          2533
UN          2448
CM          1937
OG          1033
PO           592
DM           549
CH           146
WC           135
OT            95
MP            79
SI            55
FR             1
Name: payer_code, dtype: int64

In [39]:
final['weight'].value_counts()

missing      98569
[75-100)      1336
[50-75)        897
[100-125)      625
[125-150)      145
[25-50)         97
[0-25)          48
[150-175)       35
[175-200)       11
>200             3
Name: weight, dtype: int64

#### 24 drug columns have no null/missing values

#### admission_type_id, discharge_disposition_id, admission_source_id were numerically encoded as ordinal but they are nominal data. They were handled by replacing the numbers with respective categories to enable one hot encoding.

#### Other columns have no nulls or missing vals

#### Columns having missing values or ? or unknowns are 

1. Race - 2273 missing vals have been replaced by mode - caucasian

2. Gender - 3 unknown vals were replaced with mode - Female

3. payer code - 40% vals are missing - no clear mode

4. medical speciality - 50% vals are missing - no clear mode

5. Weight - 97% vals are missing

6. diag1 - have been handled as a new category - missing

7. diag2 - have been handled as a new category - missing

8. diag3 - have been handled as a new category - missing

### Dropping Weight, payer_code and medical_speciality columns because there is no clear majority for mode imputation.

In [40]:
final=final.drop(['weight','payer_code','medical_specialty'],axis=1)

In [41]:
final.shape

(101766, 47)

In [42]:
final.columns

Index(['patient_nbr', 'diag_1', 'diag_2', 'diag_3', 'number_diagnoses',
       'admission_type_id', 'discharge_disposition_id', 'admission_source_id',
       'time_in_hospital', 'encounter_id', 'race', 'gender', 'age',
       'num_lab_procedures', 'num_procedures', 'num_medications',
       'number_outpatient', 'number_emergency', 'number_inpatient',
       'max_glu_serum', 'A1Cresult', 'metformin', 'repaglinide', 'nateglinide',
       'chlorpropamide', 'glimepiride', 'acetohexamide', 'glipizide',
       'glyburide', 'tolbutamide', 'pioglitazone', 'rosiglitazone', 'acarbose',
       'miglitol', 'troglitazone', 'tolazamide', 'examide', 'citoglipton',
       'insulin', 'glyburide-metformin', 'glipizide-metformin',
       'glimepiride-pioglitazone', 'metformin-rosiglitazone',
       'metformin-pioglitazone', 'change', 'diabetesMed', 'readmitted'],
      dtype='object')

In [43]:
final['admission_source_id'].value_counts()

EmergencyRoom                                        57494
PhysicianReferral                                    29565
NULL                                                  6781
Transferfromahospital                                 3187
Transferfromanotherhealthcarefacility                 2264
ClinicReferral                                        1104
 TransferfromaSkilledNursingFacility(SNF)              855
HMOReferral                                            187
NotMapped                                              161
NotAvailable                                           125
Court/LawEnforcement                                    16
Transferfromhospitalinpt/samefacresultinasepclaim       12
Transferfromcritialaccesshospital                        8
ExtramuralBirth                                          2
TransferfromAmbulatorySurgeryCenter                      2
NormalDelivery                                           2
SickBaby                                                

#### Combining multiple similar categories into one 

final['admission_source_id'].replace({'NULL':'NA','NotMapped':'NA','NotAvailable':'NA','PhysicianReferral':'Referral',
                                      'ClinicReferral':'Referral','HMOReferral':'Referral','Transferfromahospital':'Transfer'
                                     ,'Transferfromanotherhealthcarefacility':'Transfer',
                                      ' TransferfromaSkilledNursingFacility(SNF)':'Transfer',
                                      'Transferfromhospitalinpt/samefacresultinasepclaim':'Transfer',
                                     'Transferfromcritialaccesshospital':'Transfer','TransferfromAmbulatorySurgeryCenter':'Transfer'
                                     },inplace=True)

In [44]:
final['admission_source_id'].value_counts()

EmergencyRoom                                        57494
PhysicianReferral                                    29565
NULL                                                  6781
Transferfromahospital                                 3187
Transferfromanotherhealthcarefacility                 2264
ClinicReferral                                        1104
 TransferfromaSkilledNursingFacility(SNF)              855
HMOReferral                                            187
NotMapped                                              161
NotAvailable                                           125
Court/LawEnforcement                                    16
Transferfromhospitalinpt/samefacresultinasepclaim       12
Transferfromcritialaccesshospital                        8
ExtramuralBirth                                          2
TransferfromAmbulatorySurgeryCenter                      2
NormalDelivery                                           2
SickBaby                                                

In [45]:
final['admission_type_id'].value_counts()

Emergency        53990
Elective         18869
Urgent           18480
NULL              5291
NA                4785
Not_Mapped         320
Trauma_Center       21
Newborn             10
Name: admission_type_id, dtype: int64

#### Combining multiple similar categories into one 

In [46]:
final['admission_type_id'].replace({'NULL':'NA','NA':'NA','Not_Mapped':'NA',
                                     },inplace=True)

In [47]:
final['admission_type_id'].value_counts()

Emergency        53990
Elective         18869
Urgent           18480
NA               10396
Trauma_Center       21
Newborn             10
Name: admission_type_id, dtype: int64

In [48]:
final['discharge_disposition_id'].value_counts()

ToHome                                                  60234
ToSNF                                                   13954
ToHomeWithHomeHealthCare                                12902
NULL                                                     3691
ToShortTermHosp                                          2128
ToRehab                                                  1993
Expired                                                  1642
ToInpatientCareInst                                      1184
Not_Mapped                                                989
ToICF                                                     815
LeftAMA                                                   623
ToLongTermCare                                            412
Hospice/Home                                              399
Hospice/Med.Facility                                      372
ToPsychiatricHosp.                                        139
ToHomeIVProvider                                          108
Medicare

final['discharge_disposition_id'].replace(['ToShortTermHosp','ToSNF','ToICF','ToInpatientCareInst',
                                          'InpatientInSameHosp','AnotherHospNeonatalCare','StillPatientOROutpatient',
                                          'Hospice/Home','Hospice/Med.Facility','MedicareApprovedSwingBed',
                                   'AnotherInst.Outpatient','SameInst.Outpatient','ToRehab','ToLongTermCare',
                                          'ToNursingFac(MedicaidCertified,MedicareNotCertified)','ToFederalHealthCareFac.',
                                          'ToPsychiatricHosp.','ToCAHosp','ToUnknownHCHosp.'],
                                         ['To_further_treatment','To_further_treatment','To_further_treatment',
                                          'To_further_treatment','To_further_treatment','To_further_treatment',
                                          'To_further_treatment','To_further_treatment','To_further_treatment',
                                          'To_further_treatment','To_further_treatment','To_further_treatment',
                                          'To_further_treatment','To_further_treatment','To_further_treatment',
                                          'To_further_treatment','To_further_treatment','To_further_treatment','To_further_treatment'],inplace=True)
final['discharge_disposition_id'].replace(['NULL','Not_Mapped'],['NA','NA'],inplace=True)
final['discharge_disposition_id'].replace(['ToHomeWithHomeHealthCare','ToHomeIVProvider'],['ToHome','ToHome'],inplace=True)
final['discharge_disposition_id'].replace(['ExpiredAtHome.MedicaidOnly','ExpiredInMedFacility.MedicaidOnly'],['Expired','Expired'],inplace=True)

In [49]:
final['discharge_disposition_id'].value_counts()

ToHome                                                  60234
ToSNF                                                   13954
ToHomeWithHomeHealthCare                                12902
NULL                                                     3691
ToShortTermHosp                                          2128
ToRehab                                                  1993
Expired                                                  1642
ToInpatientCareInst                                      1184
Not_Mapped                                                989
ToICF                                                     815
LeftAMA                                                   623
ToLongTermCare                                            412
Hospice/Home                                              399
Hospice/Med.Facility                                      372
ToPsychiatricHosp.                                        139
ToHomeIVProvider                                          108
Medicare

In [50]:
#final['discharge_disposition_id'].replace({'To_further_treatment':'To_further_care'},inplace=True)

In [51]:
final['discharge_disposition_id'].value_counts()

ToHome                                                  60234
ToSNF                                                   13954
ToHomeWithHomeHealthCare                                12902
NULL                                                     3691
ToShortTermHosp                                          2128
ToRehab                                                  1993
Expired                                                  1642
ToInpatientCareInst                                      1184
Not_Mapped                                                989
ToICF                                                     815
LeftAMA                                                   623
ToLongTermCare                                            412
Hospice/Home                                              399
Hospice/Med.Facility                                      372
ToPsychiatricHosp.                                        139
ToHomeIVProvider                                          108
Medicare

##### Taking data only for DiabetesMed == Yes. This indicates that the patients have been given diabetes medication i.e. they have diabetes. This data consists of 78363 rows.

##### This is justified by DiabetesMed == No has all 24 diabetes drugs as NO. This means that they are not diagnosed with diabetes.

In [52]:
medyes = final[final['diabetesMed']=='Yes']
medyes.shape

(78363, 47)

In [53]:
copymedyes=medyes.copy()

In [54]:
copymedyes.columns

Index(['patient_nbr', 'diag_1', 'diag_2', 'diag_3', 'number_diagnoses',
       'admission_type_id', 'discharge_disposition_id', 'admission_source_id',
       'time_in_hospital', 'encounter_id', 'race', 'gender', 'age',
       'num_lab_procedures', 'num_procedures', 'num_medications',
       'number_outpatient', 'number_emergency', 'number_inpatient',
       'max_glu_serum', 'A1Cresult', 'metformin', 'repaglinide', 'nateglinide',
       'chlorpropamide', 'glimepiride', 'acetohexamide', 'glipizide',
       'glyburide', 'tolbutamide', 'pioglitazone', 'rosiglitazone', 'acarbose',
       'miglitol', 'troglitazone', 'tolazamide', 'examide', 'citoglipton',
       'insulin', 'glyburide-metformin', 'glipizide-metformin',
       'glimepiride-pioglitazone', 'metformin-rosiglitazone',
       'metformin-pioglitazone', 'change', 'diabetesMed', 'readmitted'],
      dtype='object')

In [55]:
copymedyes.iloc[:,21:44]=copymedyes.iloc[:,21:44].replace(['No','Up','Down','Steady'],[0,1,1,1])

In [56]:
copymedyes.head(15)

patient_nbr diag_1  diag_2 diag_3  number_diagnoses admission_type_id  \
1      55629189    276  250.01    255                 9         Emergency   
2      86047875    648     250    V27                 6         Emergency   
3      82442376      8  250.43    403                 7         Emergency   
4      42519267    197     157    250                 5         Emergency   
5      82637451    414     411    250                 9            Urgent   
6      84259809    414     411    V45                 7          Elective   
7     114882984    428     492    250                 8         Emergency   
8      48330783    398     427     38                 8            Urgent   
9      63555939    434     198    486                 8          Elective   
10     89869032  250.7     403    996                 9         Emergency   
11     77391171    157     288    197                 7            Urgent   
12     85504905    428  250.43  250.6                 8         Emergency   
13     77586282    428     411    427                 8         Emergency   
14     49726791    518     998    627                 8          Elective   
15     86328819    999     507    996                 9         Emergency   

    discharge_disposition_id    admission_source_id  time_in_hospital  \
1                     ToHome          EmergencyRoom                 3   
2                     ToHome          EmergencyRoom                 2   
3                     ToHome          EmergencyRoom                 2   
4                     ToHome          EmergencyRoom                 1   
5                     ToHome         ClinicReferral                 3   
6                     ToHome         ClinicReferral                 4   
7                     ToHome          EmergencyRoom                 5   
8                     ToHome  Transferfromahospital                13   
9                      ToSNF  Transferfromahospital                12   
10                    ToHome          EmergencyRoom                 9   
11                    ToHome  Transferfromahospital                 7   
12                     ToSNF          EmergencyRoom                 7   
13  ToHomeWithHomeHealthCare          EmergencyRoom                10   
14                    ToHome         ClinicReferral                 1   
15                     ToSNF          EmergencyRoom                12   

    encounter_id    ...     citoglipton insulin glyburide-metformin  \
1         149190    ...               0       1                   0   
2          64410    ...               0       0                   0   
3         500364    ...               0       1                   0   
4          16680    ...               0       1                   0   
5          35754    ...               0       1                   0   
6          55842    ...               0       1                   0   
7          63768    ...               0       0                   0   
8          12522    ...               0       1                   0   
9          15738    ...               0       1                   0   
10         28236    ...               0       1                   0   
11         36900    ...               0       1                   0   
12         40926    ...               0       1                   0   
13         42570    ...               0       1                   0   
14         62256    ...               0       1                   0   
15         73578    ...               0       1                   0   

    glipizide-metformin  glimepiride-pioglitazone  metformin-rosiglitazone  \
1                     0                         0                        0   
2                     0                         0                        0   
3                     0                         0                        0   
4                     0                         0                        0   
5                     0                         0                        0 

In [57]:
# counting the total no. of drugs given to each diabetic patient
copymedyes['Total_drugs'] =copymedyes.iloc[:,21:44].sum(axis=1)
copymedyes.head(15)

patient_nbr diag_1  diag_2 diag_3  number_diagnoses admission_type_id  \
1      55629189    276  250.01    255                 9         Emergency   
2      86047875    648     250    V27                 6         Emergency   
3      82442376      8  250.43    403                 7         Emergency   
4      42519267    197     157    250                 5         Emergency   
5      82637451    414     411    250                 9            Urgent   
6      84259809    414     411    V45                 7          Elective   
7     114882984    428     492    250                 8         Emergency   
8      48330783    398     427     38                 8            Urgent   
9      63555939    434     198    486                 8          Elective   
10     89869032  250.7     403    996                 9         Emergency   
11     77391171    157     288    197                 7            Urgent   
12     85504905    428  250.43  250.6                 8         Emergency   
13     77586282    428     411    427                 8         Emergency   
14     49726791    518     998    627                 8          Elective   
15     86328819    999     507    996                 9         Emergency   

    discharge_disposition_id    admission_source_id  time_in_hospital  \
1                     ToHome          EmergencyRoom                 3   
2                     ToHome          EmergencyRoom                 2   
3                     ToHome          EmergencyRoom                 2   
4                     ToHome          EmergencyRoom                 1   
5                     ToHome         ClinicReferral                 3   
6                     ToHome         ClinicReferral                 4   
7                     ToHome          EmergencyRoom                 5   
8                     ToHome  Transferfromahospital                13   
9                      ToSNF  Transferfromahospital                12   
10                    ToHome          EmergencyRoom                 9   
11                    ToHome  Transferfromahospital                 7   
12                     ToSNF          EmergencyRoom                 7   
13  ToHomeWithHomeHealthCare          EmergencyRoom                10   
14                    ToHome         ClinicReferral                 1   
15                     ToSNF          EmergencyRoom                12   

    encounter_id     ...     insulin glyburide-metformin glipizide-metformin  \
1         149190     ...           1                   0                   0   
2          64410     ...           0                   0                   0   
3         500364     ...           1                   0                   0   
4          16680     ...           1                   0                   0   
5          35754     ...           1                   0                   0   
6          55842     ...           1                   0                   0   
7          63768     ...           0                   0                   0   
8          12522     ...           1                   0                   0   
9          15738     ...           1                   0                   0   
10         28236     ...           1                   0                   0   
11         36900     ...           1                   0                   0   
12         40926     ...           1                   0                   0   
13         42570     ...           1                   0                   0   
14         62256     ...           1                   0                   0   
15         73578     ...           1                   0                   0   

    glimepiride-pioglitazone  metformin-rosiglitazone  metformin-pioglitazone  \
1                          0                        0                       0   
2                          0                        0                       0   
3                          0                        0                       0 

In [58]:
copymedyes['Total_drugs'].value_counts()

1    47314
2    21873
3     7778
4     1335
5       58
6        5
Name: Total_drugs, dtype: int64

In [59]:
z=copymedyes[copymedyes['Total_drugs']==1]
z.shape

(47314, 48)

In [60]:
z.iloc[:,21:44].columns.shape # 23 drugs for data having total_drugs = 1

(23,)

In [61]:
z.iloc[:,21:44].sum(axis=0) # value counts of SOLO DRUGS where total_drugs=1 

metformin                    4484
repaglinide                   384
nateglinide                   153
chlorpropamide                 59
glimepiride                  1348
acetohexamide                   0
glipizide                    3868
glyburide                    3374
tolbutamide                    13
pioglitazone                 1336
rosiglitazone                1057
acarbose                       38
miglitol                        9
troglitazone                    0
tolazamide                     20
examide                         0
citoglipton                     0
insulin                     30965
glyburide-metformin           204
glipizide-metformin             2
glimepiride-pioglitazone        0
metformin-rosiglitazone         0
metformin-pioglitazone          0
dtype: int64

In [62]:
lst = z.iloc[:,21:44].sum(axis=0).index
lst

Index(['metformin', 'repaglinide', 'nateglinide', 'chlorpropamide',
       'glimepiride', 'acetohexamide', 'glipizide', 'glyburide', 'tolbutamide',
       'pioglitazone', 'rosiglitazone', 'acarbose', 'miglitol', 'troglitazone',
       'tolazamide', 'examide', 'citoglipton', 'insulin',
       'glyburide-metformin', 'glipizide-metformin',
       'glimepiride-pioglitazone', 'metformin-rosiglitazone',
       'metformin-pioglitazone'],
      dtype='object')

In [63]:
def soloinsulin(df):
    lst=[]
    for i in range(len(df)):
        if df.iloc[i,47] >=2:
            lst.append(0)
        else:
            if df.iloc[i,38] ==1:
                lst.append(1)
            else:
                lst.append(0)
    return lst
            

In [64]:
copymedyes['Solo_Insulin'] =soloinsulin(copymedyes)
copymedyes['Solo_Insulin'].value_counts()

0    47398
1    30965
Name: Solo_Insulin, dtype: int64

In [65]:
copymedyes.columns

Index(['patient_nbr', 'diag_1', 'diag_2', 'diag_3', 'number_diagnoses',
       'admission_type_id', 'discharge_disposition_id', 'admission_source_id',
       'time_in_hospital', 'encounter_id', 'race', 'gender', 'age',
       'num_lab_procedures', 'num_procedures', 'num_medications',
       'number_outpatient', 'number_emergency', 'number_inpatient',
       'max_glu_serum', 'A1Cresult', 'metformin', 'repaglinide', 'nateglinide',
       'chlorpropamide', 'glimepiride', 'acetohexamide', 'glipizide',
       'glyburide', 'tolbutamide', 'pioglitazone', 'rosiglitazone', 'acarbose',
       'miglitol', 'troglitazone', 'tolazamide', 'examide', 'citoglipton',
       'insulin', 'glyburide-metformin', 'glipizide-metformin',
       'glimepiride-pioglitazone', 'metformin-rosiglitazone',
       'metformin-pioglitazone', 'change', 'diabetesMed', 'readmitted',
       'Total_drugs', 'Solo_Insulin'],
      dtype='object')

In [66]:
totaldrugs=copymedyes['Total_drugs']
soloinsulin=copymedyes['Solo_Insulin']

In [67]:
medyes['Total_drugs'] = totaldrugs
medyes['Solo_Insulin'] = soloinsulin
medyes.head(15)

C:\Users\user\Anaconda3\lib\site-packages\ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  """Entry point for launching an IPython kernel.
C:\Users\user\Anaconda3\lib\site-packages\ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  


patient_nbr diag_1  diag_2 diag_3  number_diagnoses admission_type_id  \
1      55629189    276  250.01    255                 9         Emergency   
2      86047875    648     250    V27                 6         Emergency   
3      82442376      8  250.43    403                 7         Emergency   
4      42519267    197     157    250                 5         Emergency   
5      82637451    414     411    250                 9            Urgent   
6      84259809    414     411    V45                 7          Elective   
7     114882984    428     492    250                 8         Emergency   
8      48330783    398     427     38                 8            Urgent   
9      63555939    434     198    486                 8          Elective   
10     89869032  250.7     403    996                 9         Emergency   
11     77391171    157     288    197                 7            Urgent   
12     85504905    428  250.43  250.6                 8         Emergency   
13     77586282    428     411    427                 8         Emergency   
14     49726791    518     998    627                 8          Elective   
15     86328819    999     507    996                 9         Emergency   

    discharge_disposition_id    admission_source_id  time_in_hospital  \
1                     ToHome          EmergencyRoom                 3   
2                     ToHome          EmergencyRoom                 2   
3                     ToHome          EmergencyRoom                 2   
4                     ToHome          EmergencyRoom                 1   
5                     ToHome         ClinicReferral                 3   
6                     ToHome         ClinicReferral                 4   
7                     ToHome          EmergencyRoom                 5   
8                     ToHome  Transferfromahospital                13   
9                      ToSNF  Transferfromahospital                12   
10                    ToHome          EmergencyRoom                 9   
11                    ToHome  Transferfromahospital                 7   
12                     ToSNF          EmergencyRoom                 7   
13  ToHomeWithHomeHealthCare          EmergencyRoom                10   
14                    ToHome         ClinicReferral                 1   
15                     ToSNF          EmergencyRoom                12   

    encounter_id     ...      glyburide-metformin glipizide-metformin  \
1         149190     ...                       No                  No   
2          64410     ...                       No                  No   
3         500364     ...                       No                  No   
4          16680     ...                       No                  No   
5          35754     ...                       No                  No   
6          55842     ...                       No                  No   
7          63768     ...                       No                  No   
8          12522     ...                       No                  No   
9          15738     ...                       No                  No   
10         28236     ...                       No                  No   
11         36900     ...                       No                  No   
12         40926     ...                       No                  No   
13         42570     ...                       No                  No   
14         62256     ...                       No                  No   
15         73578     ...                       No                  No   

   glimepiride-pioglitazone  metformin-rosiglitazone  metformin-pioglitazone  \
1                        No                       No                      No   
2                        No                       No                      No   
3                        No                       No                      No   
4                        No                       No                      No   
5                        No      

In [68]:
medyes['Total_drugs'].value_counts()

1    47314
2    21873
3     7778
4     1335
5       58
6        5
Name: Total_drugs, dtype: int64

In [69]:
medyes['Solo_Insulin'].value_counts()

0    47398
1    30965
Name: Solo_Insulin, dtype: int64

In [70]:
medyes.to_csv('Medyes.csv',index=0)

### Data with DiabetesMed=Yes, Solo_Insulin=1

In [71]:
soloins=medyes[medyes['Solo_Insulin']==1] 
soloins.shape

(30965, 49)

In [72]:
soloins['patient_nbr'].nunique() # unique patients who were treated with only insulin

23013

### Number of unique patients in soloins = 23013

In [73]:
def count_readmitted(x):
    if len(x) == 1:
        return 0
    else:
        return 1

In [74]:
insulin_effectiveness=pd.DataFrame(soloins.groupby(soloins['patient_nbr']).apply(count_readmitted),columns=['Readmitted'])

In [75]:
insulin_effectiveness.shape

(23013, 1)

In [76]:
insulin_effectiveness['Readmitted'].value_counts()

0    18636
1     4377
Name: Readmitted, dtype: int64

### Out of 23013 unique patients who were given only insulin to treat their diabetes, 4377 patients got readmitted and 18636 did not get readmitted

## 81% of diabetes patients who were given solo insulin treatment were not readmitted

## This is a good indicator that solo insulin procedure works effectively

In [77]:
temp=soloins['change'].value_counts()
len(temp.index)

2

In [78]:
def count_change(x):
    temp=x['change'].value_counts()
    if len(temp.index) == 2:
        return temp['Ch']
    else:
        return 0

In [79]:
number_change = pd.DataFrame(soloins.groupby(soloins['patient_nbr']).apply(count_change),columns=['Change_Count'])

In [80]:
number_change.shape

(23013, 1)

In [81]:
number_change['Change_Count'].value_counts()

0     21291
1      1089
2       344
3       146
4        61
5        35
6        14
7        11
8         7
9         4
12        3
11        2
10        2
20        2
14        1
19        1
Name: Change_Count, dtype: int64

#### Medication was not changed for 92% unique diabetes patients who were treated with solo insulin

### Dataset for EDA in Tableau

In [82]:
medyes.shape

(78363, 49)

In [83]:
sorted(medyes['patient_nbr'].unique())

[135,
 729,
 774,
 927,
 1152,
 1305,
 1314,
 1629,
 2025,
 3069,
 5220,
 5337,
 6174,
 6309,
 6696,
 6723,
 6939,
 8208,
 8460,
 8748,
 10377,
 11394,
 11511,
 12105,
 12240,
 12447,
 13005,
 13041,
 13320,
 13590,
 14481,
 14499,
 14535,
 14688,
 14796,
 14859,
 14949,
 15210,
 15642,
 16425,
 16623,
 18045,
 20232,
 20295,
 20565,
 20772,
 20997,
 22158,
 22527,
 24030,
 24525,
 24696,
 25362,
 25434,
 25956,
 26352,
 27333,
 27486,
 27936,
 28530,
 29961,
 30051,
 30735,
 31320,
 31392,
 31752,
 31878,
 32022,
 32337,
 32967,
 33264,
 33354,
 33930,
 34353,
 35271,
 36432,
 36738,
 36810,
 37125,
 37305,
 37863,
 38601,
 38637,
 38970,
 39006,
 39051,
 40104,
 40158,
 40392,
 40797,
 41328,
 42111,
 42120,
 42417,
 42867,
 43515,
 44154,
 44766,
 45036,
 45144,
 48150,
 48573,
 48996,
 49860,
 50472,
 52605,
 53505,
 54198,
 55629,
 55638,
 56295,
 56430,
 56520,
 58302,
 58374,
 60399,
 62487,
 62937,
 62964,
 63090,
 63288,
 63630,
 63702,
 64773,
 66618,
 66915,
 67275,
 67428,


In [84]:
new_data=pd.DataFrame(sorted(medyes['patient_nbr'].unique()),columns=['patient_nbr'])
new_data.head()

patient_nbr
0          135
1          729
2          774
3          927
4         1152

In [85]:
medyes.index=np.arange(len(medyes))

In [86]:
def race_name(x):
    lst=x['race'].values
    return lst[0]

In [87]:
new_data['race'] =medyes.groupby(medyes['patient_nbr']).apply(race_name).values
new_data.head(10)

patient_nbr             race
0          135        Caucasian
1          729        Caucasian
2          774        Caucasian
3          927  AfricanAmerican
4         1152  AfricanAmerican
5         1305        Caucasian
6         1314        Caucasian
7         1629        Caucasian
8         2025        Caucasian
9         3069        Caucasian

In [88]:
new_data['patient_nbr'].isnull().sum()

0

In [89]:
def gender_name(x):
    lst=x['gender'].values
    return lst[0]

In [90]:
new_data['gender'] = medyes.groupby(medyes['patient_nbr']).apply(gender_name).values
new_data.head(10)

patient_nbr             race  gender
0          135        Caucasian  Female
1          729        Caucasian  Female
2          774        Caucasian  Female
3          927  AfricanAmerican  Female
4         1152  AfricanAmerican  Female
5         1305        Caucasian  Female
6         1314        Caucasian  Female
7         1629        Caucasian    Male
8         2025        Caucasian  Female
9         3069        Caucasian    Male

In [91]:
new_data['gender'].value_counts()

Female    29607
Male      26571
Name: gender, dtype: int64

In [92]:
def age_cat(x):
    lst=x['age'].values
    return lst[0]

In [93]:
new_data['age']=medyes.groupby(medyes['patient_nbr']).apply(age_cat).values
new_data.head(10)

patient_nbr             race  gender      age
0          135        Caucasian  Female  [50-60)
1          729        Caucasian  Female  [80-90)
2          774        Caucasian  Female  [80-90)
3          927  AfricanAmerican  Female  [30-40)
4         1152  AfricanAmerican  Female  [50-60)
5         1305        Caucasian  Female  [60-70)
6         1314        Caucasian  Female  [40-50)
7         1629        Caucasian    Male  [70-80)
8         2025        Caucasian  Female  [70-80)
9         3069        Caucasian    Male  [60-70)

In [94]:
def readmitted(x):
    if len(x)>1:
        return 'Readmitted'
    else:
        return 'Not Readmitted'

In [95]:
new_data['readmitted_status'] =medyes.groupby(medyes['patient_nbr']).apply(readmitted).values
new_data.head()

patient_nbr             race  gender      age readmitted_status
0          135        Caucasian  Female  [50-60)        Readmitted
1          729        Caucasian  Female  [80-90)    Not Readmitted
2          774        Caucasian  Female  [80-90)    Not Readmitted
3          927  AfricanAmerican  Female  [30-40)    Not Readmitted
4         1152  AfricanAmerican  Female  [50-60)        Readmitted

In [96]:
new_data['readmitted_status'].value_counts()

Not Readmitted    43596
Readmitted        12582
Name: readmitted_status, dtype: int64

In [97]:
medyes.shape

(78363, 49)

In [98]:
def drug_type(x):
    solo=0
    oth=0
    for i in range(len(x)):
        if x.iloc[i,48]==1:
            solo+=1
        else:
            oth+=1
    if oth==0:
        return 'SoloInsulin'
    elif solo==0:
        return 'Other'
    else:
        return 'Both'

In [99]:
new_data['solo/other/both'] =medyes.groupby(medyes['patient_nbr']).apply(drug_type).values
new_data.head(10)

patient_nbr             race  gender      age readmitted_status  \
0          135        Caucasian  Female  [50-60)        Readmitted   
1          729        Caucasian  Female  [80-90)    Not Readmitted   
2          774        Caucasian  Female  [80-90)    Not Readmitted   
3          927  AfricanAmerican  Female  [30-40)    Not Readmitted   
4         1152  AfricanAmerican  Female  [50-60)        Readmitted   
5         1305        Caucasian  Female  [60-70)    Not Readmitted   
6         1314        Caucasian  Female  [40-50)        Readmitted   
7         1629        Caucasian    Male  [70-80)    Not Readmitted   
8         2025        Caucasian  Female  [70-80)    Not Readmitted   
9         3069        Caucasian    Male  [60-70)    Not Readmitted   

  solo/other/both  
0           Other  
1           Other  
2           Other  
3           Other  
4     SoloInsulin  
5     SoloInsulin  
6            Both  
7     SoloInsulin  
8           Other  
9     SoloInsulin

In [100]:
medyes[medyes['patient_nbr']==1314]

patient_nbr diag_1 diag_2  diag_3  number_diagnoses admission_type_id  \
11620         1314    996    411     401                 9            Urgent   
14480         1314    414    411  250.01                 7         Emergency   
14590         1314     78    250     414                 7         Emergency   

      discharge_disposition_id admission_source_id  time_in_hospital  \
11620                   ToHome   PhysicianReferral                 2   
14480                   ToHome       EmergencyRoom                 1   
14590                   ToHome       EmergencyRoom                 3   

       encounter_id     ...      glyburide-metformin glipizide-metformin  \
11620      60254142     ...                       No                  No   
14480      70190028     ...                       No                  No   
14590      70601076     ...                       No                  No   

      glimepiride-pioglitazone  metformin-rosiglitazone  \
11620                       No                       No   
14480                       No                       No   
14590                       No                       No   

       metformin-pioglitazone  change  diabetesMed  readmitted  Total_drugs  \
11620                      No      No          Yes         >30            1   
14480                      No      No          Yes         <30            1   
14590                      No      Ch          Yes          NO            2   

      Solo_Insulin  
11620            1  
14480            1  
14590            0  

[3 rows x 49 columns]

In [101]:
def diagnosis(x):
    for i in range(len(x)):
         if(( x.iloc[i,1].startswith('250.') == True) or ( x.iloc[i,2].startswith('250.') == True) or 
            ( x.iloc[i,3].startswith('250.') == True)):
                

SyntaxError: unexpected EOF while parsing (<ipython-input-101-951c532d502f>, line 5)

In [1]:
import pandas as pd

In [2]:
final=pd.read_csv('Final_Data.csv')

In [13]:
final[final['diag_1']=='?']

518      ?
1006     ?
1267     ?
1488     ?
3197     ?
7845     ?
14503    ?
19714    ?
32514    ?
37693    ?
49516    ?
56931    ?
57058    ?
57737    ?
60314    ?
61369    ?
61763    ?
66848    ?
86018    ?
87181    ?
98396    ?
Name: diag_1, dtype: object

In [ ]:
]